### PR TITLE
Refactor to use validation module

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- ~Documentation Added~
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR does three things:
- move the utility-method `make_index()` to the utils module
- move the utility-method `_exclude_on_fail()` to a new validation module
- add tests to show that `require_data()` implements logical-and, i.e., all combinations across dimensions are required, (see #768

### CODE REVIEW

File: tests/test_feature_validation.py [hunk 1]
```diff
@@ -3,43 +3,61 @@
...
+    columns=IAMC_IDX + TEST_YEARS,
+)
 
 
 @pytest.mark.parametrize(
     "kwargs",
     (
         dict(),
         dict(variable="Primary Energy"),
+        dict(variable="Primary Energy", year=[2005, 2010]),
```
Comment (phackstock) [inline tests/test_feature_validation.py:26]: Is there a specific reason you're using `dict()` over `{}`?
```suggestion
        {"variable": "Primary Energy", "year": [2005, 2010]},
```

File: pyam/validation.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import logging
+import pandas as pd
+
+from pyam.utils import make_index, s
```
Comment (phackstock) [inline pyam/validation.py:4]: I'd suggest renaming the function `s` into something more descriptive.

**Review by phackstock — COMMENTED** [review]
Looks good to me. Two small comments below.

File: tests/test_feature_validation.py (hunk 1 above)
Comment (danielhuppmann) [reply tests/test_feature_validation.py:26]: I think readability is better thanks to the syntax-highlighting...

File: pyam/validation.py (hunk 2 above)
Comment (phackstock) [reply pyam/validation.py:4]: Maybe `plural_s`? Although I typed that before I'd seen the use in a docstring. There it's quite readable, so ok from my side to keep it this way if we don't have any good ideas.

File: tests/test_feature_validation.py (hunk 1 above)
Comment (phackstock) [reply tests/test_feature_validation.py:26]: Fair

**Review by phackstock — APPROVED** [review]
Good to be merged from my side. 
If you find a good alternative name for the `s` function you can change it otherwise it's fine as well to keep it as is and merge.

**Pull request merged**